src/auth/policy/resolver.go

```diff
@@ -1,41 +1,349 @@
+			}
+		} else {
+			// action is an exact name (e.g. SignTransaction)
+
+			// lookup for effect already associated to this action
+			effect, ok := rslvr.exactActions[action]
+			if ok {
+				// we already had an effect for prefix so we acculmulate the statement
+				err := effect.addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+			} else {
+				// we create the effect and insert it into the map of exact actions
+				newEffect := new(effectResolver)
+				err := newEffect.addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+				rslvr.exactActions[action] = newEffect
+			}
+		}
+	}
+
+	return nil
+}
+
+func (rslvr *actionsResolver) isAuthorized(op *Operation) (res *result) {
+	res = &result{
+		op: op,
+	}
+
+	// do we have an exact match for the action?
+	actionEffects, hasExact := rslvr.exactActions[op.Action]
+	if hasExact {
+		// if we have an exact match and it resolves to Deny we return Deny
+		res = actionEffects.isAuthorized(op)
+		if res.deny != nil {
+			return res
+		}
+	}
+
+	// do we have a prefix matching the action?
+	actionPrefix, _, hasPrefix := rslvr.prefixedActions.LongestPrefix(op.Action)
+	if !hasPrefix {
+		// no prefix so we return
+		return res
+	}
+
+	// we have a prefix
+	// we walk all matching prefix starting from root up to the longest matching prefix
+	// if we meet a single Deny we stop walking and return a Deny
+	rslvr.prefixedActions.WalkPath(actionPrefix, func(_ string, v interface{}) bool {
+		r := v.(*effectResolver).isAuthorized(op)
+		if r.deny != nil {
+			// we met a Deny
+			res = r
+			return true
+		}
+
+		if res.allow == nil {
+			res = r
+		}
+
+		return false
+	})
+
+	return
 }
 
-func NewResolver(policies []types.Policy) (*Resolver, error) {
-	return &Resolver{}, nil
+// effectResolver is associated to each pair path/action (possibly prefixes)
+type effectResolver struct {
+	allow *statement
+	deny  *statement
+}
+
+func (rslvr *effectResolver) isAuthorized(op *Operation) (res *result) {
+	return &result{
+		op:    op,
+		deny:  rslvr.deny,
+		allow: rslvr.allow,
+	}
+}
+
+// addStatement cumulates the effect of multiple statement
+// if a singe statement is Deny, resolver will always resolve to Deny
+// if multiple Deny are passed only the first one is kept
+func (rslvr *effectResolver) addStatement(sttmnt *statement) error {
+	if rslvr.deny != nil {
+		return nil
+	}
+
+	switch sttmnt.Effect {
+	case "Allow":
+		rslvr.allow = sttmnt
+		return nil
+	case "Deny":
+		rslvr.deny = sttmnt
+		rslvr.allow = nil
+		return nil
+	default:
+		return fmt.Errorf("invalid effect %q", sttmnt.Effect)
+	}
+}
+
+// RadixResolver allows to perform authorization checks
+
+// It is built from a set of Policy by aggregating all statements into some
+// radix trees structure allowing optimized authorization checks
+type RadixResolver struct {
+	// exactPath maps an exact path to an action resolver
+	exactPath map[string]*actionsResolver
```

Comment:
As far as I see we could we don't need `exactPath` and it can use radix.Tree to manage every scenario

Reply:
Hmmm...

First, I am not sure how we could handle cases where you have statements with both the exact path and prefix path (like `/store/keys*` and `/store/keys`).

Second, 
- storing in radix is more expensive in storage than a regular map (because we have intermediary nodes), we accept this extra cost to accelerate prefix pattern resolution
- for an exact path, a normal map is less expensive in storage and enables faster queries than a radix

So having both a radix for prefixes and a map for exact sounds the best option (btw this is also how it is implemented in Vault).

Reply:
In terms of performance we are comparing a log(n) with O1, what IMO does not pay off the complexity of having two separate flows to maintain in the code and the additional complexity that means. I cannot see strong reasons to not use just radix for both alternatives, it might be worthy to have a chat about it to explore how it could be done.